agent.go

```diff
@@ -385,6 +385,100 @@ func (c *controller) agentClose() {
 	agent.networkDB.Close()
 }
 
+type Task struct {
+	Name       string
+	EndpointID string
+	EndpointIP string
+	Info       string
+}
+
+type ServiceInfo struct {
+	VIP   string
+	Tasks []Task
+	Ports []string
+}
+type epRecord struct {
+	ep   EndpointRecord
+	info string
+}
+
+func (n *network) Services() map[string]ServiceInfo {
+	eps := make(map[string]epRecord)
+
+	if !n.isClusterEligible() {
+		return nil
+	}
+	agent := n.getController().getAgent()
+	if agent == nil {
+		return nil
+	}
+
+	// Walk through the libnetwork's "endpoint_table" and fetch the driver
+	// agnostic endpoint info
+	entries := agent.networkDB.GetTableByNetwork("endpoint_table", n.id)
+	for eid, value := range entries {
+		var epRec EndpointRecord
+		if err := proto.Unmarshal(value.([]byte), &epRec); err != nil {
+			logrus.Errorf("Unmarshal of %s failed", eid)
+			continue
+		}
+		eps[eid] = epRecord{
+			ep: epRec,
+		}
+	}
+
+	// Walk through the driver's tables, have the driver decode the entries
+	// and return the tuple {ep ID, value}. value is a string that coveys
+	// relevant info about the endpoint.
+	d, err := n.driver(true)
+	if err != nil {
+		logrus.Errorf("Can't get the driver for network, %s\n", n.ID())
+		return nil
+	}
+	for _, table := range n.driverTables {
+		entries := agent.networkDB.GetTableByNetwork(table, n.id)
+		for key, value := range entries {
+			epID, info := d.DecodeTableEntry(table, key, value.([]byte))
+			if ep, ok := eps[epID]; !ok {
+				logrus.Errorf("epRecord update attempt for driver info, but missing for %s", epID)
+			} else {
+				ep.info = info
+				eps[epID] = ep
+			}
+		}
+	}
+
+	// group the endpoints into a map keyed by the service name
+	sinfo := make(map[string]ServiceInfo)
+	for ep, epr := range eps {
+		var (
+			s  ServiceInfo
+			ok bool
+		)
+		if s, ok = sinfo[epr.ep.ServiceName]; !ok {
+			s = ServiceInfo{
+				VIP: epr.ep.VirtualIP,
+			}
+		}
+		ports := []string{}
+		if s.Ports == nil {
+			for _, port := range epr.ep.IngressPorts {
+				p := fmt.Sprintf("Target %d, Publish %d", port.TargetPort, port.PublishedPort)
```

Comment:
I would separate `Target` and `Publish` labels from their value with a semicolon.